[Sol->Yul] Populating internal dispatch on demand

cameel opened the pull request:
Resolves #6788. Adds a mechanism necessary to implement code generation for internal library function calls via pointers in #8485.

~This is not a perfect solution because internal functions that are called directly rather than assigned to variables are still included in the internal dispatch. That's because generator treats both of these cases the same way, by storing function ID in a variable, and in the case of a direct call the variable is simply not used. I would have to detect that which is not easy and probably not even desirable. So I did not implement it and only functions that are never referenced at all will be skipped. It's still an improvement and it's the minimum needed for #8485.~ Edit: the mechanism has been changed and now functions that are called directly are detected and not included in the internal dispatch.

~The PR is not finished yet. Stuff that's missing:~
- [x] ~Docstrings for all new functions~
- [x] ~`semanticTests/viaYul/array_function_pointers` fails. Turns out that internal dispatch should be generated even if it's empty because calling an empty function pointer is a runtime error, not a compile-time one.~
- [x] ~I did not check the case of function pointers being initialized in constructor (as pointed out in the issue), but I suspect it's already handled.~

Status: **working on requested changes**
- [x] Get rid of excessive docs and reword in some places.
- [x] Minor changes to helpers and `internalDispatch()` in `IRGenerationContext`.
- [x] Generate dispatch functions even if they're not called - to simplify implementation.
- [x] Use `visit(FunctionCall)` to avoid having to add and remove candidates for dispatch.
- [x] ~Put dispatch generation in a separate class if it's still too complex after changes above.~

cameel commented:
The PR is now complete and ready for review.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -153,10 +153,70 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
+		});
+	}
+
+	string funName = internalDispatchFunctionName(_arity);
```
Since the function is only generated once, it is very important that the `_functions` argument is correct. If the `internalDispatch` function is called again, it will not re-generate the function. Is there any way to ensure that?

cameel replied:
Sure. I'll add an assert that the dispatch function is not already in the function collector.

cameel replied:
Done.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -153,10 +153,70 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
+	}
+
+	string funName = internalDispatchFunctionName(_arity);
+	return m_functions.createFunction(funName, [&, cases(move(cases))]() {
```
Maybe pull the loop over `_functions` inside the lambda to avoid the moving here?

cameel replied:
ok

cameel replied:
Actually I think it's better to leave it as is. If I put this code in the lambda I don't have to move `cases` but then I have to move or copy `_functions`. Or risk having a dangling reference if the list I'm referencing goes out of scope. Or the list could be modified from the outside even if it does not go out of scope.

BTW, this part is now in #8943.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -153,10 +155,135 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
+	solAssert(m_dispatchableInternalFunctionReferences.count(&_expression) == 0, "");
+
+	m_dispatchableInternalFunctionReferences[&_expression] = &_function;
+	return internalDispatchFunctionName(functionArity(_function));
```
I don't see any use of the return value.

cameel replied:
Right. This is a leftover from before one of the rewrites when I tried to use an earlier version of this function in `endVisit(FunctionCall)` where the name would have been used. I'll remove it.

In general I don't really like that these functions in the context are returning function names anyway. I'd prefer them to be separate calls but I'm just following a convention used in the context.

cameel replied:
Done.

chriseth commented on libsolidity/codegen/ir/IRGenerator.h:
```diff
@@ -62,9 +62,14 @@ class IRGenerator
 	/// The resulting code is stored in the function collector in IRGenerationContext.
-	void generateQueuedFunctions();
+	/// @return The list of internal dispatch candidates that should be passed to the context.
+	/// This needs to be done only for candidates from the deployment code since they may end up
```
```suggestion
	/// This needs to be done only for candidates from the creation code since they may end up
```

cameel replied:
Done.

chriseth commented on libsolidity/codegen/ir/IRGenerator.h:
```diff
@@ -62,9 +62,14 @@ class IRGenerator
+	/// @return The list of internal dispatch candidates that should be passed to the context.
+	/// This needs to be done only for candidates from the deployment code since they may end up
+	/// in storage variables and be called from the runtime code. Candidates detected in the
+	/// runtime code of the contract can be ignored.
```
can be ignored by who?

cameel replied:
By `generate()`. Right, this goes a bit too deep into internal details of how this function is currently being used by `IRGenerator`. I'll remove this part.

cameel replied:
Removed. I removed or shortened most of the docstrings.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.h:
```diff
@@ -154,6 +252,29 @@ class IRGenerationContext
+	/// Collection of functions that need to be callable via internal dispatch.
+	/// These are internal functions which satisfy all of the following conditions:
+	/// 1. Are referenced by name in an expression other than a direct function call.
+	/// 2. There exists at least one call of any internal function of the same arity via a pointer.
```
Do need that condition during code creation? Such a dispatch routine should be removed by the optimizer.

cameel replied:
I didn't want to assume that it will be removed by the optimizer. But if I can make this assumption then yeah, just generating dispatches corresponding to all detected functions will make the logic much simpler.

cameel replied:
Done. This simplified the code a lot.

cameel commented:
Moving it out of draft status. Not ready for another review yet though :) Still working on it.

cameel commented:
- [ ] test

cameel commented:
Ready for another round of review. I think I fixed all the issues that were pointed out here and on gitter. The implementation is now much simpler. I have also extracted all the refactoring changes into a separate PR (#8943).

cameel commented:
This is now based on #8952 which is the last in a series of PRs which replaced #8943.

Looks like the tests need to be retriggered. I'll be rebasing it when the PRs below get merged anyway so the tests will run then.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1465,7 +1480,10 @@ void IRGeneratorForStatements::endVisit(MemberAccess const& _memberAccess)
 					if (auto const* function = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration))
+					{
 						define(_memberAccess) << to_string(function->id()) << "\n";
+						m_context.internalFunctionCalled(_memberAccess);
```
Should this rather be `internalFunctionAccessed`?

cameel replied:
Right. I'll change it.

cameel replied:
Done.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -121,22 +123,51 @@ string IRGenerationContext::newYulVariable()
+
+	if (m_directInternalFunctionCalls.count(&_expression) == 0)
+	{
+		FunctionDefinition const& resolvedFunction = IRHelpers::referencedFunctionDeclaration(_expression)->resolveVirtual(mostDerivedContract());
```
Are these always virtual calls?

cameel replied:
Right. I'll change it.

cameel replied:
Good catch. They are not. But I'm not sure how it is passing tests then. I found this problem earlier and it was failing the test that calls a contract method from the base class and I was sure I fixed it. I need to take a closer look at what I did there.

cameel replied:
I know what happened. I had two similar problems with virtual resolution, one affecting only direct calls and the other only calls via pointers. I caught the former in tests because that's a more common case. The one pointed out here was not covered.

I fixed my code and also added a test that would detect such a situation (`inherited_function_via_pointer.sol`).

cameel committed:
Generate internal dispatch only for functions that might actually get called via pointers

- This also adds support for internal library calls as a side-effect since they'll now be pulled into the internal dispatch automatically.

cameel commented:
I had to rewrite the code in `IRGenerator` and `IRGenerationContext` to match changes requested in https://github.com/ethereum/solidity/pull/8951#discussion_r427633806.

cameel commented:
Regarding https://github.com/ethereum/solidity/pull/8951#discussion_r427633806, I think I found a neater way to organize this code. I moved `generateInternalDispatchFunctions()` to `IRGenerator`. I think it's the right place for it. Now context prepares the dispatch map and the generator does all the generating.

cameel commented:
Also, this PR is now again on top of #8952.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -679,9 +693,10 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		else
 		{
 			YulArity arity = YulArity::fromType(*functionType);
+			m_context.internalFunctionCalledViaPointer(arity);
```
Maybe `throughDispatch` instead of `viaPointer`?

cameel replied:
Done.

Also changed `via_pointer` to `through_dispatch` in test names.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -121,52 +123,53 @@ string IRGenerationContext::newYulVariable()
+	}
+}
+
+void IRGenerationContext::internalFunctionCalledViaPointer(YulArity const& _arity)
```
In which situations do we need this function and cannot just rely on `internalFunctionCalledDirectly` not being called and then the function being added to the dispatch via `internalFunctionAccessed`?

cameel replied:
Trying to call something via an uninitialized pointer. In this case `internalFunctionAccessed()` and `internalFunctionCalledDirectly()` never get called for this particular function. If there are no other functions of this arity, the corresponding dispatch function does not get generated.

[calling_uninitialized_function.sol](https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/functionCall/calling_uninitialized_function.sol):
```
Invalid IR generated:
:167:32: Error: Function not found.
                let expr_17 := dispatch_internal_in_1_out_1(expr_15_functionIdentifier, _4)
                ^--------------------------^
:167:17: Error: Variable count mismatch: 1 variables and 0 values.
                let expr_17 := dispatch_internal_in_1_out_1(expr_15_functionIdentifier, _4)
                ^-------------------------------------------------------------------------^
```

chriseth replied:
Ah, now I understand! This basically says "we need a dispatch function for this expression" while the others say "add this function to the switch inside the dispatch function" right?

cameel replied:
Exactly :)

That's why I initially named these functions referring to what they do (register dispatch, add candidate, etc.). Naming them from the "outside" perspective is valid too but hides details like these.

chriseth replied:
Yeah ok, you are right :)

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1733,7 +1751,14 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
+
+		solAssert(resolvedFunctionDef.functionType(true), "");
+		solAssert(resolvedFunctionDef.functionType(true)->kind() == FunctionType::Kind::Internal, "");
+		m_context.internalFunctionAccessed(_identifier);
```
Why is it not a problem that you mark the function accessed via the expression (and `internalFunctionAccessed` then using `.referencedDeclaration`) but store the value after doing virtual function lookup?

cameel replied:
Because `internalFunctionAccessed()` gets `_identifer` (and extracts the function definition on its own), rather than `resolvedFunctionDef`. I'm using `resolvedFunctionDef` here only to check assertions.

I think I was using `resolvedFunctionDef` (possibly incorrectly) in one of the previous versions here so it made more sense back then but yeah, now it's misleading. I'll change the assertions to use `functionDef` instead.

cameel replied:
Done. Assertions are now using `functionDef`.

chriseth replied:
I'm still confused. You assign the correct function id to the variable, so it will call the right function, but inside `internalFunctionAccessed` ( https://github.com/ethereum/solidity/pull/8797/files#diff-f841efc1479453da2e05b159a7f1da6bR165 ) you store the base function in the dispatch map and request code generation for the base function.

Shouldn't this go wrong in the following case:
```
contract A {
  function f() virtual internal {}
  function g() internal { (f)(); }
}
contract B is A {
  function f() override internal {} // code for this is never generated
  function h() public { g(); }
}
```

chriseth replied:
Edited the above to force call via pointer.

cameel replied:
You're right it does not work with your example. I fixed it now by rewriting `internalFunctionAccessed()` to accept a function definiton as well. That was how it was supposed to work back when I was still adding and removing candidates. I got this confused during one of the rewrites when I decided that I could get by just with the expression.

I added an example similar to yours as a test (`internal_virtual_function_calls_through_dispatch.sol`). Not exactly the same since the `(f)()` syntax makes an unrelated assertion fail, as I explained on gitter. This assertion is already in develop, so I'll be fixing it separately.

Anyway, this is another bug in this code that slipped through and I think I just need to start adding more tests. There are lots of tests already and I assumed that they cover most of the corner cases so I was adding them sparingly. But looks like I need a better coverage. It's hard to spot these errors myself when rewriting because of the assumptions I have from the previous iterations...

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.h:
```diff
@@ -102,7 +104,13 @@ class IRGenerationContext
+	InternalDispatchMap consumeInternalDispatchMap();
+	bool internalDispatchClean() const { return m_internalDispatchMap.empty() && m_directInternalFunctionCalls.empty(); }
+
+	void internalFunctionCalledThroughDispatch(YulArity const& _arity);
```
Can you add some documentation to these three function? 1-2 lines should be enough.

cameel replied:
Sure.

cameel replied:
Done.

cameel replied:
A bit longer (3-4 lines) but to shorten it I would have to cut the first sentence that says what the function does. It's is a bit redundant but still, the docstring feels a bit weird without it.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -121,52 +123,53 @@ string IRGenerationContext::newYulVariable()
+		solAssert(functionType, "");
+
+		YulArity arity = YulArity::fromType(*functionType);
+		m_internalDispatchMap.try_emplace(arity);
```
Is this line needed?

cameel replied:
Oh, I thought that `[]` would not initialize the new element but looks like it does.
Yeah, it's unnecessary after all.

chriseth commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -164,6 +174,71 @@ void IRGenerator::generateQueuedFunctions()
+		string funName = IRNames::internalDispatch(arity);
+		m_context.functionCollector().createFunction(funName, [&]() {
+			Whiskers templ(R"(
+				function <functionName>(fun <comma> <in>) <arrow> <out> {
```
```suggestion
				function <functionName>(fun<?+in>, <in></+in>) <?+out>-> <out> </+out>{
```

cameel replied:
This is something that's already in `develop` and this PR does not change it (only moves the function) so I fixed it in a separate PR: #9026.

chriseth commented on libsolidity/codegen/ir/IRGenerator.cpp:
```diff
@@ -164,6 +174,71 @@ void IRGenerator::generateQueuedFunctions()
+					<#cases>
+					case <funID>
+					{
+						<out> <assignment_op> <name>(<in>)
```
```suggestion
						<+?out><out> := </+out><name>(<in>)
```

cameel replied:
This is something that's already in `develop` and this PR does not change it (only moves the function) so I fixed it in a separate PR: #9026.

chriseth commented on libsolidity/codegen/ir/IRGenerationContext.cpp:
```diff
@@ -121,49 +123,55 @@ string IRGenerationContext::newYulVariable()
+{
+	solAssert(
+		IRHelpers::referencedFunctionDeclaration(_expression) &&
+		_function.resolveVirtual(mostDerivedContract()) ==
```
Wouldn't it be better to check that `_function` is either exactly the referenced function declaration or the virtually-resolved one?

chriseth replied:
Maybe it also does not matter too much, this is not very far from where it is called.

cameel replied:
That's what I had here at first but I changed it because this was shorter and still correct.

chriseth commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -583,6 +583,20 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!functionType->bound() &&
```
Does this mean that all bound functions go through the dispatch routine?

cameel replied:
No. They just don't have to be added to the list of direct calls because `endVisit(MemberAccess)` short-circuits for bound function calls and won't call `internalFunctionAccessed()` which would add them to the dispatch.
